### PR TITLE
API: Fix listSSHKeyPairs API when listing all resources (listall=true & projectid=-1)

### PR DESCRIPTION
**Pull request by Pearl1594**
### Description

This PR fixes the issue noticed when listing all ssh key pairs including ones owned by projects. Exception:
```
(primateqa) 🐱 > list sshkeypairs listall=true projectid=-1
🙈 Error: (HTTP 530, error code 4250) DB Exception on: com.mysql.cj.jdbc.ClientPreparedStatement: SELECT ssh_keypairs.id, ssh_keypairs.uuid, ssh_keypairs.account_id, ssh_keypairs.domain_id, ssh_keypairs.keypair_name, ssh_keypairs.fingerprint, ssh_keypairs.public_key FROM ssh_keypairs  INNER JOIN account ON ssh_keypairs.account_id=account.id  WH ORDER BY ssh_keypairs.id DESC  LIMIT 0, 500
```

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [X] Minor
- [ ] Trivial

#### Post Fix:
```
(localcloud) 🐱 > list sshkeypairs listall=true page=1 pagesize=20 filter=name,
{
  "count": 1,
  "sshkeypair": [
    {
      "name": "ssh"
    }
  ]
}
(localcloud) 🐱 > list sshkeypairs listall=true projectid=-1 page=1 pagesize=20 filter=name,
{
  "count": 2,
  "sshkeypair": [
    {
      "name": "test"
    },
    {
      "name": "ssh"
    }
  ]
}
```

### CODE REVIEW

File: framework/db/src/main/java/com/cloud/utils/db/GenericDaoBase.java [hunk 1]
```diff
@@ -1293,7 +1294,8 @@ protected void addJoins(StringBuilder str, Collection<JoinBuilder<SearchCriteria
             fromIndex += onClause.length();
         }
 
-        str.delete(str.length() - 4, str.length());
+        int diff =  str.toString().trim().toLowerCase(Locale.ROOT).endsWith("where") ? 6 : 4;
+        str.delete(str.length() - diff, str.length());
```
Comment (rohityadavcloud) [inline framework/db/src/main/java/com/cloud/utils/db/GenericDaoBase.java:1298]: @Pearl1594 is there any unit test to check/cover this, could this cause a regression? Maybe explore a more localised fix that's limited to ssh-key based DaoImpl?

File: framework/db/src/main/java/com/cloud/utils/db/GenericDaoBase.java [hunk 2]
```diff
@@ -1293,7 +1294,8 @@ protected void addJoins(StringBuilder str, Collection<JoinBuilder<SearchCriteria
             fromIndex += onClause.length();
         }
 
-        str.delete(str.length() - 4, str.length());
+        int diff =  str.toString().trim().toLowerCase(Locale.ROOT).endsWith("where") ? 6 : 4;
```
Comment (DaanHoogland) [inline framework/db/src/main/java/com/cloud/utils/db/GenericDaoBase.java:1297]: @Pearl1594 I'd leave out the trim() it is not clear why the 6 or 4 like this (or add a comment, but I prefer the code to explain itself)

Comment (sureshanaparti) [reply framework/db/src/main/java/com/cloud/utils/db/GenericDaoBase.java:1297]: +1, add comment if code is not self-explanatory.

**Comment by weizhouapache** [comment]
@Pearl1594 
I suggest to append 'WHERE' only when it is required, like code changes below
```
@@ -1262,10 +1262,11 @@ public abstract class GenericDaoBase<T, ID extends Serializable> extends Compone
 
     @DB()
     protected void addJoins(StringBuilder str, Collection<JoinBuilder<SearchCriteria<?>>> joins) {
+        boolean hasWhereClause = true;
         int fromIndex = str.lastIndexOf("WHERE");
         if (fromIndex == -1) {
             fromIndex = str.length();
-            str.append(" WHERE ");
+            hasWhereClause = false;
         } else {
             str.append(" AND ");
         }
@@ -1287,13 +1288,20 @@ public abstract class GenericDaoBase<T, ID extends Serializable> extends Compone
             .append(" ");
             str.insert(fromIndex, onClause);
             String whereClause = join.getT().getWhereClause();
-            if ((whereClause != null) && !"".equals(whereClause)) {
+            if (!StringUtils.isEmpty(whereClause)) {
+                if (!hasWhereClause) {
+                    str.append(" WHERE ");
+                    hasWhereClause = true;
+                }
                 str.append(" (").append(whereClause).append(") AND");
             }
             fromIndex += onClause.length();
         }
 
-        str.delete(str.length() - 4, str.length());
+        if (hasWhereClause) {
+            // Remove ' AND' OR 'AND ' at the end of line
+            str.delete(str.length() - 4, str.length());
+        }
 
         for (JoinBuilder<SearchCriteria<?>> join : joins) {
             if (join.getT().getJoins() != null) {
```

**Comment by rohityadavcloud** [comment]
This looks like a bugfix, should it be considered for 4.16.1 @sureshanaparti @nvazquez @DaanHoogland @weizhouapache

**Comment by nvazquez** [comment]
Thanks @rohityadavcloud I agree, @sureshanaparti please include it on 4.16.1 if necessary

**Comment by sureshanaparti** [comment]
@Pearl1594 Is this PR good to go in 4.16.1? If so, please address outstanding comment and rebase with 4.16. Thanks.

**Pull request closed without merging**